pkg/exporter/exporter.go

```diff
@@ -2,38 +2,80 @@ package exporter
 
 import (
 	"fmt"
-	"github.com/swatisehgal/resource-topology-exporter/pkg/apis/topocontroller/v1alpha1"
 	"log"
+	"os"
+
+	"github.com/davecgh/go-spew/spew"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+
+	v1alpha1 "github.com/swatisehgal/resource-topology-exporter/pkg/apis/topocontroller/v1alpha1"
+	clientset "github.com/swatisehgal/resource-topology-exporter/pkg/generated/clientset/versioned"
 )
 
-type crdExporter struct {
-	NodeResources []v1alpha1.NUMANodeResource
+type CRDExporter struct {
+	cli      *clientset.Clientset
+	hostname string
 }
 
-type CRDExporter interface {
-	Run() error
-}
+func NewExporter() (*CRDExporter, error) {
+	clientConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("Please run from inside the cluster")
+	}
 
-func NewExporter(instance []v1alpha1.NUMANodeResource) *crdExporter {
-	return &crdExporter{
-		NodeResources: instance,
+	cli, err := clientset.NewForConfig(clientConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Error building example clientset: %s", err.Error())
 	}
-}
 
-func (e *crdExporter) Run() error {
-	err := e.Update()
+	hostname, err := os.Hostname()
 	if err != nil {
-		return fmt.Errorf("Unable to update Exporter: %v", err)
+		return nil, fmt.Errorf("Failed to get Hostname:%v", err)
 	}
-	return nil
+
+	return &CRDExporter{
+		cli:      cli,
+		hostname: hostname,
+	}, nil
 }
 
-func (e *crdExporter) Update() error {
-	log.Printf("Exporter Update called NodeResources is: %+v", e.NodeResources)
+func (e *CRDExporter) CreateOrUpdate(namespace string, resources []v1alpha1.NUMANodeResource) error {
+	log.Printf("Exporter Update called NodeResources is: %+v", resources)
+
+	nrt, err := e.cli.TopocontrollerV1alpha1().NodeResourceTopologies(namespace).Get(e.hostname, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		nrtNew := v1alpha1.NodeResourceTopology{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: e.hostname,
+			},
+			Nodes: resources,
+		}
+
+		nrtCreated, err := e.cli.TopocontrollerV1alpha1().NodeResourceTopologies(namespace).Create(&nrtNew, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("Failed to create v1alpha1.NodeResourceTopology!:%v", err)
+		}
+		log.Printf("CRD instance created resTopo: %v", spew.Sdump(nrtCreated))
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
 
-	err := UpdateCustomResourceDefinition(e.NodeResources)
+	nrtMutated := v1alpha1.NodeResourceTopology{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            e.hostname,
+			ResourceVersion: nrt.ObjectMeta.ResourceVersion,
```

Comment:
I'm not sold we need this, need to review the update flow. Up until now I'm just doing what the old code did.

Reply:
Not providing the resourceVersion was causing the following error `metadata.resourceVersion: Invalid value: 0x0: must be specified for an update`. I came across [this](https://github.com/abonas/kubeclient/issues/403) thread and implemented by getting the object, extracting resourceVersion from it and providing that in the update. I am open to an alternative way of achieving this. I am going to approve the PR for now and we can work on an alternative way of implementing this in a separate PR.